Proposal: Allow narrowing ORM fields

I am currently working on implementing [phantom types for Python](https://github.com/antonagestam/phantom-types), which, when implemented throughout an application, would allow statically proving that a value saved on model is validated.

To be able to use this with django-stubs there would need to be some way of narrowing the types that ORM fields hold.

One example would be a field for country codes:

```python
from phantom.ext.iso3166 import CountryCode


class Member(models.Model):
    name = CharField()
    country: CountryCode = CharField(max_length=2)
```

This makes mypy error if I try to assign a string value to the field:

```python
m = Member()
m.country = "SE"  # E: Incompatible types in assignment (expression has type "str", variable has type "CountryCode")
```

Unless I first prove that the string is a valid country code:

```python
country_code = "SE"
assert isinstance(country_code, CountryCode)
m.country = country_code
# ... or
m.country = CountryCode.parse(country_code)
```

Unfortunately mypy raises an error when annotating ORM fields:

```
Incompatible types in assignment (expression has type "CharField[Union[str, int, Combinable, None], Optional[str]]", variable has type "CountryCode") [assignment]
```

Although my usecase is phantom types, I believe a feature that enables this would also work equally well with `typing.NewType` types.

This can be worked around with using the usual dirty tricks of ignore comments or using e.g. `typing.cast(CountryCode, CharField())`. But it would be very nice to see builtin support for this. I think it would be optimal if this builtin support also verifies that the annotated type is a subtype of the type that the field holds, so that the type annotated to a `CharField` must be a subtype of `str` and so on.

What do you think?

_PS.: This is a proposal to get a discussion started, I'm not asking anyone to implement this for me._

@sobolevn Cool, I did not know about that! I think I'm running into a variance issue though. 

I setup a minimal Django project to test with, and testing with this definition just to have less moving parts to think about (also tested with `NewType` with same results):

```python
from __future__ import annotations

from django.db import models


class CountryCode(str):
    pass


class Member(models.Model):
    name = models.CharField(max_length=200)
    country: models.Field[CountryCode, CountryCode] = models.CharField(max_length=2)
```

I get:

```
phango/models.py:12: error: Incompatible types in assignment (expression has type "CharField[Union[str, int, Combinable], str]", variable has type "Field[CountryCode, CountryCode]") [assignment]
        country: models.Field[CountryCode, CountryCode] = models.CharField(max_length=2)
                                                          ^
```

If I change the definitions of `_ST` and `_GT` that `Field` is generic with regards to, to be contravariant I seem to get the expected behavior without errors. I can open a PR introducing this change.

One further issue though, that might be unrelated to this at all is that I don't get type errors passing field values when instantiating the model. I'm assuming this is unrelated because it passes even with this example:

```python
from __future__ import annotations

from django.db import models


class Member(models.Model):
    value: models.Field[int, int] = models.IntegerField()


m = Member(value="test")  # Expected error here
```

--- a/django-stubs/db/models/fields/__init__.pyi
+++ b/django-stubs/db/models/fields/__init__.pyi
@@ -38,9 +38,9 @@ _ErrorMessagesToOverride = Dict[str, Any]
 
 _T = TypeVar("_T", bound="Field")
 # __set__ value type
-_ST = TypeVar("_ST")
+_ST = TypeVar("_ST", contravariant=True)
 # __get__ return type
-_GT = TypeVar("_GT")
+_GT = TypeVar("_GT", covariant=True)
 
 class Field(RegisterLookupMixin, Generic[_ST, _GT]):
     _pyi_private_set_type: Any

--- a/tests/typecheck/fields/test_base.yml
+++ b/tests/typecheck/fields/test_base.yml
@@ -163,3 +163,5 @@
         book.published = 2006  # E: Incompatible types in assignment (expression has type "int", variable has type "Year")
         book.published = Year(2006)
         reveal_type(book.published)  # N: Revealed type is 'main.Year*'
+        def accepts_int(arg: int) -> None: ...
+        accepts_int(book.published)

--- a/tests/typecheck/fields/test_base.yml
+++ b/tests/typecheck/fields/test_base.yml
@@ -151,3 +151,15 @@
 
                 class MyModel(AuthMixin, models.Model):
                     pass
+-   case: can_narrow_field_type
+    main: |
+        from typing import cast, NewType
+        from django.db import models
+        Year = NewType("Year", int)
+        class Book(models.Model):
+            published = cast(models.Field[Year, Year], models.IntegerField())
+        book = Book()
+        reveal_type(book.published)  # N: Revealed type is 'main.Year*'
+        book.published = 2006  # E: Incompatible types in assignment (expression has type "int", variable has type "Year")
+        book.published = Year(2006)
+        reveal_type(book.published)  # N: Revealed type is 'main.Year*'

